### DISCUSSION TITLE
Object as a single state

### DISCUSSION THREAD

**Discussion by majidsajadi** [pmndrs/zustand#1444] [Q&A]
Hello guys

I need to store a set of states which the type (schema) is same. for example i have a list of items which each has a name `itemName` and some settings assign to it which can be manipulated by users. and users can add more items.

```ts
 item1: {
  color: 'red',
  content: "red content"
},
item2: {
  color: 'blue',
  content: "blue content"
},
```

i need to store this settings in my global state. the current implementation is like this:

```ts
export type RootState = DataSlice;

export interface IBuildQueryParams {
  color?: string;
  content?: string;
}


export interface DataSlice {
  settings: Record<string, IBuildQueryParams>;
  setSettings: (name: string, setting: Partial<IBuildQueryParams>) => void;
}

const defaultState = {
  settings: {},
};

export const createDataSlice: StateCreator<
  RootState,
  [["zustand/persist", unknown]],
  [],
  DataSlice
> = (set) => ({
  ...defaultState,
  set: (name: string, setting: Partial<IBuildQueryParams>) =>
    set((state) => {
      return {
        ...state,
        queries: {
          ...state.queries,
          [name]: {
            ...state.queries[name],
            ...setting,
          },
        },
      };
    }),
});
```

and in the component i get settings for each item like this:

```ts
const [settings, setSettings] = useAppStore(
  (state) => [state.settings, state.setSettings],
  shallow
);

if (itemName) {
 const itemColor= settings[itemName]?.color
}

```

my questions are:

what are the alternative ways. using Object as state is not desirable since im getting the whole state in component which may cause not desired rerender.
also `itemName` may be undefined which cause a lot if nullish checking before getting state. and manipulating the current state is hard too. for example changing color from a specific item setting state.

im thinking using `Map` but since `Map` in ES is not mutable i guess its would be a bad choice. also thinking about `immutable.js` which i dont like to add another library for this specific reason.

any suggestions? am i missing any point?

**Comment by dai-shi** [comment]
Hi, I'm not exactly sure what is your concern,

> using Object as state is not desirable since im getting the whole state in component which may cause not desired rerender.

but if you want to avoid re-renders without manual selectors, [react-tracked](https://react-tracked.js.org) may help.
